FullStorySegmentMiddleware/Classes/FullStoryMiddleware.m

```diff
@@ -125,102 +127,148 @@ - (SEGPayload *) getNewPayloadWithFSURL:(SEGContext * _Nonnull)context {
 
-- (NSString *) getSuffixStringFromSimpleObject: (NSObject *) obj{
-    NSString * suffix;
-    // defualt to string
-    if([obj isKindOfClass:[NSNumber class]]){
+- (NSString *) getSuffixStringFromSimpleObject: (NSObject *) obj {
+    // default to no suffix;
+    NSString * suffix = @"";
+    if ([obj isKindOfClass:[NSNumber class]]) {
         // defaut to real
         suffix = @"_real";
         NSNumber *n = (NSNumber *) obj;
         const char *type = n.objCType;
-        if([@"i" isEqualToString:@(type)]) suffix = @"_int";
-        // bool type number doesn't get encoded into 'B' but check anyway
-        else if ([@"B" isEqualToString:@(type)])suffix = @"_bool";
-    }else if ([obj isKindOfClass:[NSDate class]]){
+        if([@"i" isEqualToString:@(type)]) {
+            suffix = @"_int";
+        } else if ([@"B" isEqualToString:@(type)]) {
+            // bool-type gets encoded as number number, and doesn't get encoded into 'B', but check anyway
+            suffix = @"_bool";
+        }
+    } else if ([obj isKindOfClass:[NSDate class]]) {
         suffix = @"_date";
-    }else if([obj isKindOfClass:[NSString class]]){
+    } else if([obj isKindOfClass:[NSString class]]) {
         suffix = @"_str";
     }
+
     return suffix;
 }
 
-- (NSDictionary *) getObjectFromArrayObject: (NSArray *) arr withKey:(NSString *) key{
-    // FS does not accept array of dicts, only premitive arrays allowed
-    // So if it's an array of dicts, then we need to convert the root array to a dict so it becomes nested dicts
+- (NSDictionary *) getDictionaryFromArrayObject:(NSArray *) arr withKey:(NSString *) key {
+    // FS does not accept array of dicts, only "premitive" arrays allowed
+    // So if it's an array of dicts or nested arrays, then we need to convert the root array to a dict so it becomes nested dicts
     NSMutableDictionary *dict = [[NSMutableDictionary alloc] init];
 
-    for (int i = 0; i < [arr count]; i++){
+    for (int i = 0; i < [arr count]; i++) {
         NSObject *item = arr[i];
-        if([item isKindOfClass:[NSDictionary class]]){
-            // if array of dict objects, we need to convert the array to object, and give each item a key (this is a server side restriction)
-            NSMutableDictionary *tempDict = [[NSMutableDictionary alloc]initWithDictionary:@{[@"idx" stringByAppendingFormat:@"%d",i]:[self getSuffixedProps:(NSDictionary *) item]}];
-            [tempDict addEntriesFromDictionary:[dict valueForKey:key]];
-            [dict setObject:tempDict forKey:[key stringByReplacingOccurrencesOfString:@"_" withString:@""]];
-        }else if([item isKindOfClass:[NSArray class]]){
+        
+        if ([item isKindOfClass:[NSDictionary class]]) {
+            // if array of dicts, we then loop through all dicts, flatten out each key/val into arrays, we will loose the object association but it allows user to search for each key/val in the array in FS (i.e. searching for one product when array of products are sent)
```

Comment:
No longer `if`. We now know it's an array of dicts, so I would drop `if array of dicts, we then`. This comment is also rather long and should be split into multiple lines.